core/executor/src/wasmtime/trampoline.rs

```diff
@@ -0,0 +1,295 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The trampoline is the dynamically generated entry point to a runtime host call.
+//!
+//! This code is based on and large parts are copied from wasmtime's
+//! wasmtime-api/src/trampoline/func.rs.
+
+use cranelift_codegen::{Context, binemit, ir, isa};
+use cranelift_codegen::ir::{InstBuilder, StackSlotData, StackSlotKind, TrapCode};
+use cranelift_frontend::{FunctionBuilder, FunctionBuilderContext};
+use wasmtime_jit::CodeMemory;
+use wasmtime_runtime::{VMContext, VMFunctionBody};
+use wasm_interface::{HostFunctions, Function, Value, ValueType};
+use std::{cmp, mem, ptr};
+
+use crate::error::{Error, Result};
+use crate::wasmtime::function_executor::{FunctionExecutorState, FunctionExecutor};
+
+/// The top-level host state of the "env" module. This state is used by the trampoline function to
+/// construct a `FunctionExecutor` which can execute the host call.
+pub struct TrampolineState {
+	externals: &'static [&'static dyn Function],
+	trap: Option<Error>,
+	/// The executor state stored across host calls during a single Wasm runtime call.
+	/// During a runtime call, this MUST be `Some`.
+	pub executor_state: Option<FunctionExecutorState>,
+	// The code memory must be kept around on the state to prevent it from being dropped.
+	#[allow(dead_code)]
+	code_memory: CodeMemory,
+}
+
+impl TrampolineState {
+	/// Construct a new `TrampolineState` which owns the given code memory.
+	pub fn new<HF: HostFunctions>(code_memory: CodeMemory) -> Self {
+		TrampolineState {
+			externals: HF::functions(),
+			trap: None,
+			executor_state: None,
+			code_memory,
+		}
+	}
+
+	/// Resets the trap error to None and returns the current value.
+	pub fn reset_trap(&mut self) -> Option<Error> {
+		mem::replace(&mut self.trap, None)
+	}
+}
+
+/// This is called by the dynamically generated trampoline taking the function index and reference
+/// to the call arguments on the stack as arguments. Returns 0 on success and 1 on an error.
+unsafe extern "C" fn stub_fn(vmctx: *mut VMContext, func_index: u32, values_vec: *mut i64) -> u32 {
+	if let Some(state) = (*vmctx).host_state().downcast_mut::<TrampolineState>() {
+		match stub_fn_inner(
+			vmctx,
+			state.externals,
+			func_index,
+			state.executor_state.as_mut(),
+			values_vec
+		) {
+			Ok(()) => 0,
+			Err(err) => {
+				state.trap = Some(err);
+				1
+			}
+		}
+	} else {
+		// Well, we can't even set a trap message, so we'll just exit without one.
+		1
+	}
+}
+
+/// Implements most of the logic in `stub_fn` but returning a `Result` instead of an integer error
+/// for the sake of readability.
+unsafe fn stub_fn_inner(
+	vmctx: *mut VMContext,
+	externals: &[&dyn Function],
+	func_index: u32,
```

Comment:
I'll switch all of the `call_id`s to `func_index` as I think that is more descriptive. I left them as `call_id` to keep it as similar to the source code from wasmtime as possible, but I can see why it would be worth changing for clarity.